usr/lib/common/lock_sess_mgr.c

```diff
@@ -865,233 +917,338 @@ CK_RV session_mgr_set_op_state(STDLL_TokData_t *tokdata, SESSION *sess,
+                    if (!mech_param) {
+                        if (context)
+                            free(context);
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(mech_param, ptr3, ctx->mech.ulParameterLen);
+                }
+            }
+            break;
+
+        case STATE_SIGN:
+        case STATE_VERIFY:
+            {
+                SIGN_VERIFY_CONTEXT *ctx =
+                    (SIGN_VERIFY_CONTEXT *)(cur_data + sizeof(OP_STATE_DATA));
+
+                len = sizeof(SIGN_VERIFY_CONTEXT) + ctx->context_len +
+                                                    ctx->mech.ulParameterLen;
+                if (len != op_data->data_len) {
+                    TRACE_ERROR("%s\n", ock_err(ERR_SAVED_STATE_INVALID));
+                    return CKR_SAVED_STATE_INVALID;
+                }
+                if (auth_key == CK_INVALID_HANDLE) {
+                    TRACE_ERROR("%s\n", ock_err(ERR_KEY_NEEDED));
+                    return CKR_KEY_NEEDED;
+                }
+                ptr1 = (CK_BYTE *) ctx;
+                ptr2 = ptr1 + sizeof(SIGN_VERIFY_CONTEXT);
+                ptr3 = ptr2 + ctx->context_len;
+
+                if (ctx->context_len) {
+                    context = (CK_BYTE *) malloc(ctx->context_len);
+                    if (!context) {
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(context, ptr2, ctx->context_len);
+                }
+
+                if (ctx->mech.ulParameterLen) {
+                    mech_param = (CK_BYTE *) malloc(ctx->mech.ulParameterLen);
+                    if (!mech_param) {
+                        if (context)
+                            free(context);
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(mech_param, ptr3, ctx->mech.ulParameterLen);
+                }
+            }
+            break;
+
+        case STATE_DIGEST:
+            {
+                DIGEST_CONTEXT *ctx =
+                    (DIGEST_CONTEXT *)(cur_data + sizeof(OP_STATE_DATA));
+
+                len = sizeof(DIGEST_CONTEXT) + ctx->context_len +
+                                                ctx->mech.ulParameterLen;
+                if (len != op_data->data_len) {
+                    TRACE_ERROR("%s\n", ock_err(ERR_SAVED_STATE_INVALID));
+                    return CKR_SAVED_STATE_INVALID;
+                }
+                ptr1 = (CK_BYTE *) ctx;
+                ptr2 = ptr1 + sizeof(DIGEST_CONTEXT);
+                ptr3 = ptr2 + ctx->context_len;
+
+                if (ctx->context_len) {
+                    context = (CK_BYTE *) malloc(ctx->context_len);
+                    if (!context) {
```

Comment:
Where is 'context' freed in the non-error case? After the 'switch' context is set to NULL, but not freed?

Reply:
The newly allocated context is set into sess->xxx_ctx->context in the switch and thus belongs now to the session. Between allocation and the place when it is sett into the session it can't fail. Once it is set into the session, it will be freed when the operation is terminated (e.g. by a Final call, or when the session is closed).
 
So if setting another operation fails after one has been set successfully, the successful one is still set in the session, and will thus be freed when the (first) operation is terminated.

We could argue that if setting one operations fails (e.g. the second one), then all previously set operations should also be terminated. This could be done. However, the second while loop can only fail due to allocation failures (out of memory). All other possible errors have already been checked in the first while loop. The second while loop would not run at all if the first one detected an error.

Reply:
ok, sounds reasonable.